contrib_bots/lib/gcalendar.py

```diff
@@ -0,0 +1,233 @@
+
+    The primary calendar is marked as "(primary)"
+    '''
+    credentials = get_credentials()
+    http = credentials.authorize(httplib2.Http())
+    service = discovery.build('calendar', 'v3', http=http)
+
+    cal_response =  service.calendarList().list().execute();
+    calendar_list = cal_response.get('items', []);
+
+    for calendar in calendar_list:
+        name = calendar['summary']
+        cid = calendar['id']
+
+        if 'primary' in calendar:
+            name += ' (primary)'
+
+        print(name + ' - ID: ' + cid)
+
+def reply_message(msg, original_msg, client):
+    '''Replies to a specific message, through the channel of origin.
+
+    Keyword arguments:
+    msg -- content of the reply
+    original_msg -- message that is being replied
+    client -- client from the message handler
+    '''
+    message_type = original_msg['type']
+
+    if message_type == 'private':
+        original_sender = original_msg['sender_email']
+
+        client.send_message(dict(
+            type=message_type,
+            to=original_sender,
+            content=msg,
+        ))
+    elif message_type == 'stream':
+        original_stream = original_msg['display_recipient']
+        original_subject = original_msg['subject']
+
+        client.send_message(dict(
+            type=message_type,
+            to=original_stream,
+            subject=original_subject,
+            content=msg,
+        ))
+    else:
+        raise 'Unknown message type'
+
+class GCalendarHandler(object):
+    '''
+    This plugin facilitates creating Google Calendar events.
+    '''
+    def usage(self):
+        return '''
+            This plugin will allow users to create events
+            in a Google Calendar. Users should preface
+            messages with "@gcalendar"
+
+            Before running this, make sure to register a new
+            project in the Google Developers Console
+            (https://console.developers.google.com/start/api?id=calendar)
+            and download the "client secret" to ~/.gcal_client_secret.json
+            '''
+
+    def triage_message(self, message):
+        original_content = message['content']
+
+        is_gcalendar = original_content.startswith('@gcalendar')
+
+        return is_gcalendar
+
+    def handle_message(self, message, client, state_handler):
+        original_content = message['content']
+        original_sender = message['sender_email']
+
+        content = message['content'].replace('@gcalendar ', '')
+
+        # Identify event name and date (ISO 8601 notation)
+        #   Expected formats:
+        #       "Meeting with John" 2003-01-14T13:37:00+01:00 2003-01-14T15:37:00+01:00
+        #       Trip 2003-01-14
+        input_re = '(\w+|"[\S ]+") ((\d{4}-(0[1-9]|1[1-2])-(0[1-9]|1[1-9]|2[1-9]|3[0-1])T(0[0-9]|1[0-9]|2[0-3]):[0-5][0-9]:[0-5][0-9](\+|\-)(0[1-9]|1[1-4])\:[03]0 ?){2}|(\d{4}-(0[1-9]|1[1-2])-(0[1-9]|1[1-9]|2[1-9]|3[0-1])$))'
```

Comment:
I think you can use [dateparser](https://dateparser.readthedocs.io/en/latest/) to handle this.